### DISCUSSION TITLE
RDOTNET problem with wide data sets

### DISCUSSION THREAD

**Discussion by rdstern** [IDEMSInternational/R-Instat#9787] [General]
I have put the July release as a milestone, not because this issue will be solved by then.  However this is a serious problem and we should have a clear strategy and time-scale by the July launch.

I asked Ian Edwards about this issue.  His first reply was as follows:

> Initial thoughts on the crash problem:

>Recovering from crashes...
>- Periodic auto-save and/or auto-save before each called to engine.Evaluate
>- Detect if R-Instat did not shut down correctly last time and message to resume from auto-save

> Preventing crashes...
>- Can you auto-save and then dispose of the engine regularly and resume immediately with a fresh instance?
>- Do the above either periodically, or when testing shows there is a higher probability of crashes (before particular operations, or after N calls to Evaluate)

> Fixing in RDotNet...
>- Looks unlikely, the package is mostly the effort of a single developer and has not been updated since Aug 15
>- I've added a comment on the RDotNet issue to try to get the main developer involved.

> Are you still planning to progress to InstatStudio (moving away from VB.NET and VisualStudio towards the C++ and Qt technologies used by RStudio)? In which case, at least this problem will go away then.  Similarly, if Climsoft moves from VB to Python in the next year we should also not have problems with later releases.

He has a reply to his query on the RDOTNET site as follows:

> I think that looking into this issue is unlikely given the complexity of getting a debug environment configured and operational (compiled R, windbg, gdb on windows, etc...). We host the R engine with RDotNet in a distributed .net app (we offload processing to multiple instances) and have not run into this issue. My sense is that 'calling the same operation in a tight loop' just isn't a good idea.

> As a postmortem to what we've created in our shop, I wouldn't use RDotNet again. Hosting the engine in a .Net process is just incredibly problematic from a runtime and deployment perspective, not to mention the effort required to get distributed/multi-process hosting working. I just don't think that 'accessing R from .Net' is actually something desirable when you just want to pass in scripts and get the results/plots back out.

As a result Ian suggests as follows:

>I agree with the reply on the RDotNet site that it's not really viable to attempt to investigate the internal RDotNet crashing issue. They also suggest RDotNet is not the best choice going forward, and this might be why development has stopped.

>I think it should be possible to workaround the issue.

>To me it also looks like the issue could be used in a positive way for the next phase of the software:
>a) it provides a clear incentive to transfer over to more suitable language that will be well-received by the community
>b) relatively little effort was invested in VB - all of the R code is transferable, and the main work would be to recreate the dialogs and layout and apply the same logic.

>RStudio is very well written and is released under a suitable open source licence.  I would suggest building a future version of Instat directly on top of the RStudio code. It's also something I'd be very interested in being involved with.

My view is as follows:
1. The periodic saving is very important.
2. We should still look to see what we can do within our current "system" to minimise the crashes.
3. We always had Ian's long-term solution in mind.  But it was not consistent with our strategy for development, whereby most of the coding is done in Kenya.
4. The change is likely to be larger than Ian implies.  Although there are no commands in the front-end, there is a lot of code and structure.  
5. As well as Ian's solution I wonder what other languages talk to R - and more solidly than the DOTNET family.  Might we be able to move to another language more easily?  Might this then also have the advantage of being more system independent, i.e. is there an intermediate option?
6. At this stage in R-Instat we almost certainly want to move "in parallel".  So we will want to be using the current  and further 2017 versions in VB.NET, even if a possible team started soon on a conversion.

**Comment by dannyparsons** [comment]
Something to look into is JASP https://jasp-stats.org/ and it's spin off jamovi https://www.jamovi.org/ which are both front ends to R which are platform independent. I've downloaded jamovi but haven't looked into how they connect to R. If anyone has time that would be worth investigating. I see that they use some Python scripts in the code but not sure what they are used for.
